sdk/core/Azure.Core/src/Pipeline/Internal/UserAgentString.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// Information about the package to be included in UserAgent telemetry
+    /// </summary>
+    public class UserAgentString
+    {
+        private string _userAgent;
+
+        /// <summary>
+        /// Initialize an instance of <see cref="UserAgentString"/>.
+        /// </summary>
+        /// <param name="assembly">The <see cref="Assembly"/> used to generate the package name and version information for the <see cref="UserAgentString"/> value.</param>
+        /// <param name="applicationId">An optional applicationId to be prepended to the <see cref="UserAgentString"/> value. This value behaves exactly like the <see cref="DiagnosticsOptions.ApplicationId"/> property.</param>
+        public UserAgentString(Assembly assembly, string? applicationId = null)
+        {
+            _userAgent = GenerateUserAgentString(assembly, applicationId);
+        }
+
+        /// <summary>
+        /// Creates an instance of a <see cref="UserAgentString"/> based on the Type provided.
+        /// </summary>
+        /// <param name="applicationId"></param>
+        /// <typeparam name="T">The type contained by the Assembly used to generate package name and version information.</typeparam>
+        /// <returns></returns>
+        public static UserAgentString FromType<T>(string? applicationId = null)
+        {
+            var assembly = Assembly.GetAssembly(typeof(T));
```

Comment:
Should we handle null assembly here?

Reply:
If the type parameter is null, it will throw - https://github.com/microsoft/referencesource/blob/master/mscorlib/system/reflection/assembly.cs#L75

It can apparently still return null somehow, I would assume if someone hacked in a type somehow - https://github.com/microsoft/referencesource/blob/master/mscorlib/system/reflection/assembly.cs#L81